File: solutions/java/src/trafficsignalcontrolsystem/states/intersection/EastWestGreenState.java
package trafficsignalcontrolsystem.states.intersection;

import trafficsignalcontrolsystem.IntersectionController;
import trafficsignalcontrolsystem.enums.Direction;
import trafficsignalcontrolsystem.enums.LightColor;

public class EastWestGreenState implements IntersectionState {
    @Override
    public void handle(IntersectionController context) throws InterruptedException {
        System.out.printf("\n--- INTERSECTION %d: Cycle -> East-West GREEN ---\n", context.getId());

        // Turn East and West green, ensure North and South are red
        context.getLight(Direction.EAST).startGreen();
        context.getLight(Direction.WEST).startGreen();
        context.getLight(Direction.NORTH).setColor(LightColor.RED);
        context.getLight(Direction.SOUTH).setColor(LightColor.RED);

        // Wait for green light duration
        Thread.sleep(context.getGreenDuration());

        // Transition East and West to Yellow
        context.getLight(Direction.EAST).transition();
        context.getLight(Direction.WEST).transition();

        // Wait for yellow light duration
        Thread.sleep(context.getYellowDuration());

        // Transition East and West to Red
        context.getLight(Direction.EAST).transition();
        context.getLight(Direction.WEST).transition();

        // Change the intersection's state back to let North-South go
        context.setState(new NorthSouthGreenState());
    }
}

File: solutions/csharp/trafficsignalsystem/States/Intersection/EastWestGreenState.cs
class EastWestGreenState : IIntersectionState
{
    public void Handle(IntersectionController context)
    {
        Console.WriteLine($"\n--- INTERSECTION {context.GetId()}: Cycle -> East-West GREEN ---");

        // Turn East and West green, ensure North and South are red
        context.GetLight(Direction.EAST).StartGreen();
        context.GetLight(Direction.WEST).StartGreen();
        context.GetLight(Direction.NORTH).SetColor(LightColor.RED);
        context.GetLight(Direction.SOUTH).SetColor(LightColor.RED);

        // Wait for green light duration
        Thread.Sleep((int)context.GetGreenDuration());

        // Transition East and West to Yellow
        context.GetLight(Direction.EAST).Transition();
        context.GetLight(Direction.WEST).Transition();

        // Wait for yellow light duration
        Thread.Sleep((int)context.GetYellowDuration());

        // Transition East and West to Red
        context.GetLight(Direction.EAST).Transition();
        context.GetLight(Direction.WEST).Transition();

        // Change the intersection's state back to let North-South go
        context.SetState(new NorthSouthGreenState());
    }
}